lib/http_clients/salesforce_token_server.ex

```diff
@@ -0,0 +1,92 @@
+defmodule HttpClients.SalesforceTokenServer do
+  @moduledoc """
+  An agent that stores a valid Salesforce API token.
+
+  First setup it on your `application.ex`, e.g.:
+
+  ```elixir
+  config = [
+    url: "https://login.salesforce.com",
+    client_id: "some client_id",
+    client_secret: "some client_secret",
+    username: "fulano@creditas.com",
+    password: "somepassword123",
+    grant_type: "password"
+  ]
+  children = [{HttpClients.SalesforceTokenServer, name: MyApp.SalesforceTokenServer, config: config}]
+  Supervisor.start_link(children, name: MyApp.Supervisor)
+  ```
+
+  Then you can retrieve the token:
+
+  ```elixir
+  HttpClients.SalesforceTokenServer.get_token(MyApp.SalesforceTokenServer)
+  ```
+
+  Also you can update the token when needed:
+
+  ```elixir
+  HttpClients.SalesforceTokenServer.update_token(MyApp.SalesforceTokenServer)
+  ```
+
+  Or you can get a new token and set it by yourself:
+
+  ```elixir
+  config = [
+    url: "https://login.salesforce.com",
+    client_id: "some client_id",
+    client_secret: "some client_secret",
+    username: "fulano@creditas.com",
+    password: "somepassword123",
+    grant_type: "password"
+  ]
+  {:ok, token} = HttpClients.SalesforceTokenServer.request_new_token(config)
+  HttpClients.SalesforceTokenServer.set_token(MyApp.SalesforceTokenServer, token)
+  ```
+  """
+  use Agent
+  require Logger
+
+  @spec start_link(keyword()) :: {:ok, pid()} | {:error, any()} | no_return()
+  def start_link(opts) when is_list(opts) do
+    {config, opts} = Keyword.pop!(opts, :config)
+
+    Agent.start_link(
+      fn ->
+        {:ok, token} = request_new_token(config)
+        Logger.info("#{__MODULE__} started with pid #{inspect(self())}")
+        %{token: token, config: config}
+      end,
+      opts
+    )
+  end
+
+  @doc "Request an authenticated token to Salesforce"
+  @spec request_new_token(keyword()) :: {:ok, ExForce.OAuthResponse.t()} | {:error, any()}
+  def request_new_token(config) do
+    {url, opts} = Keyword.pop!(config, :url)
```

Comment:
aqui não seria:
```suggestion
    {url, config} = Keyword.pop!(config, :url)
```
?
🤔

Reply:
pode ser, eu não quis repetir config duas vezes com dados diferentes só

Reply:
o segundo elemento da tupla não é o elemento "imutado"?  daí a config é igual na real ne? acho que é semelhante à [here](https://github.com/bcredi/http-clients/pull/18/files#diff-f9abd8c7ed0fa81449d94d1714221036c58ee0495294aac32c55ec846ef55ddfR52) (linha 52) não é

Reply:
no caso o `pop/2` realmente tira o valor da Keyword.
apliquei a sugestão usando o nome usado no [ExForce](https://github.com/chulkilee/ex_force/blob/master/lib/ex_force/oauth.ex#L104) (`payload`)